### ISSUE TITLE
Zerocash paper: Does Receive always imply I can spend it?

### ISSUE THREAD

**Issue by defuse** [Electric-Coin-Company/zcash#723]
The completeness property requires that all spendable coins show up in the output of Receive:

> C outputs 1 if ... for some i ... the coin c<sub>i</sub> is not returned by Receive...

This does not imply that if Receive outputs a coin then I can spend that coin. Unless one of the other properties proves the implication, it's possible there's a vulnerability where I'm led to believe I got sent money but when I try to spend it it doesn't work.

Close this ticket once we've found where that's proven in the paper, or if we're sufficiently convinced it's true.

(An easy workaround would be to have Receive create a pour transaction that spends the coin, verify it itself, then discard it without actually sending it to the ledger).

**Comment by daira** [comment]
Assigning to myself to reread the thread where this was discussed (it's the Faerie Gold thread on zdep, but we didn't only discuss that attack).

**Comment by defuse** [comment]
It seems like there's some duplication between Receive and the NP statement. Maybe the NP statement ought to be used inside Receive.

**Comment by defuse** [comment]
btw there's an error in the paper: receive prepended the v_i in the computation of cm_i but in the two descriptions of the NP statement it's appended.

**Comment by daira** [comment]
Yes, we should define helper functions for validation and reference them both from the description of the NP statement and from Receive.

**Comment by daira** [comment]
And yes, the paper is wrong in section 4.2.

**Comment by daira** [comment]
Note that the current proposed Faerie Gold fix doesn't require any extra checks in Receive: the Pour statement already checks that the two ρ values are constructed correctly given the coin commitment check that the receiver does in step 3 (b) ii. first bullet point.
